### ISSUE TITLE
Add some way of getting the command executed

### ISSUE THREAD

**Issue by KazWolfe** [SpongePowered/SpongeAPI#1413]
## Synopsis

When a command is executed, `CommandCallable#process()` is called, but the command name itself is not passed. This is problematic, because it encourages code re-use for trivial applications that can all be handled in one master command without problem. Similarly `SendCommandEvent` does not have a way of figuring *who* executed the command, only that a command has been executed.

## Expected Behavior

Ideally, `CommandCallable#process()` should have a new field added: `name` (of type `String`). `name` will contain the value that was used to call the command. For example, if I were to call `/hi` or `/example:hi`, the `name` variable should be set to `hi`.

Sponge's Help system should (still) only include help for the primary alias.

## Reasoning for Request

This would purely be a time-saving feature that allows developers to not have to register multiple methods of a single plugin, and not write more than one command that does effectively the same thing (ie, create a `/gm` command with aliases `/creative` and `/survival`).

## Related Issues

* #676, which was closed as "unneeded" -- even if it's unneeded, it's still nice to have for developers who want to use this feature.

**Comment by Deamon5550** [comment]
> Similarly SendCommandEvent does not have a way of figuring who executed the command, only that a command has been executed.

The cause contains the source as the root element.

**Comment by ST-DDT** [comment]
I'm not sure whether there is really much code duplication due to this behaviour.

First of all you could share the same code for the different `GameMode` commands you only have to use a field+constructor to create the command for different `GameMode`s.

Your command registration may look like this then:

``` java
GameModeCommand survivalCmd = new GameModeCommand(GameModes.SURVIVAL);
GameModeCommand creativeCmd = new GameModeCommand(GameModes.CREATIVE);
commandManager.register(survivalCmd, "survival");
commandManager.register(creativeCmd, "creative");
```

(This is pseudocode, please refer to the [command  docs](https://docs.spongepowered.org/master/en/plugin/commands/index.html) for more help)

For the `/gm` command it is basically a different command because you have an extra parameter for the GameMode and thus you would have to customize the command handling for it anyway.
However with the current system you could still register the `/gm` command and reuse the above commands as subexecutors. Please refer to the [command  docs](https://docs.spongepowered.org/master/en/plugin/commands/childcommands.html) for in detail examples.

**Comment by KazWolfe** [comment]
Okay, here's an example that demonstrates more of the code re-use issue, then.

I'm writing a command to allow players to select a kit, let's call the parent command `/kit <kitname>`. However, I also need to make all of the kit names accessible as a command, so `/<kitname>`.

With the current way of using Sponge to do this, I need to register a new command for every single kit, which I can't do dynamically at all.

So, instead of just using aliases for all of my kits, I need to register each kit as a unique command, that just calls a hardcoded value (so in pseudocode):

```
Game.setKitForPlayer(player, Kit.WARRIOR);
```

While, if it were possible to get the command executed, I could just do something like this and make it much easier (with the alises registered as `Kit.getAllKitNames()`):

```
 Game.setKitForPlayer(player, Kit.getKitByName(commandName));
```

This avoids having to declare a bunch of "useless" commands, and also allows me to dynamically generate commands without having to go through some annoying steps (like defining custom help data for each command, or have each command have its own entry in `/help`).

**Comment by gravityfox** [comment]
So i think you're confused as to how commands work...

Commands have a critical problem when placed on the root level. (`/tp` and `/give` are root level commands) Namespace collisions. That is why you generally don't dynamically allocate commands at the root level.

You want to be able to dynamically bind and unbind command aliases from the root level. This is problematic in and of itself, but let's say it's okay to do it anyway.

Your solution requires your command to take up ALL OF THE REMAINING NAMESPACE, because you can't register all of your aliases up front. You don't know what kits you're registering until you load your config, and you can't change aliases for a command.

You are much better off dynamically registering a base command with a parameter, and storing the mapping.

Here's an example:

``` java
CommandManager manager = Sponge.getGame().getCommandManager();
// Make this a field in your plugin that you can access later on.
Map<String, CommandMapping> kitsCommandMappings = new Hashmap<>();
for (String kitName : Kits.getKitNames()){
    Optional<CommandMapping> mappingOpt = manager.register(this, new KitShortcutCommand(kitName), kitName);
    if (mappingOpt.isPresent()) {
        kitsCommandMappings.put(kitName, mappingOpt.get());
    } else {
        // Someone else is already using this alias, do something to handle this.
    }
}
```

Then when someone removes a kit, all you need to do is:

``` java
String kitName = "kitToRemove";
Sponge.getGame().getCommandManager().removeMapping(kitsCommandMappings.get(kitName));
```

Basically I think you're approaching the problem the wrong way. The command system was designed very carefully to work well without convoluted code. If you need to know what alias was used, then you're approaching the problem from the completely wrong direction.

**Comment by ryantheleach** [comment]
> With the current way of using Sponge to do this, I need to register a new command for every single kit, which I can't do dynamically at all.

Are you certain?

I've previously unregistered and registered commands, so unless something has changed you should still be able to do so?

**Comment by KazWolfe** [comment]
Closing as per suggestion of @gravityfox. I'd still prefer to have the exposed command name called in the command method (especially when the alias is known), but it's probably better to abandon the Bukkit way of registering commands that we were using previously.
